src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,564 @@
+      WT_BLOCK_ENDS_IN_CHUNK(chunk_off, block_off, chunk_size, block_size) ||   \
+      WT_BLOCK_MIDDLE_IN_CHUNK(chunk_off, block_off, chunk_size, block_size) || \
+      WT_BLOCK_IN_CHUNK(chunk_off, block_off, chunk_size, block_size)
+
+#define WT_CHUNK_MARK_VALID(session, chunkcache, chunk)                            \
+    if (!chunk->valid) {                                                           \
+        (void)__wt_atomic_addv32(&chunk->valid, 1);                                \
+        __wt_spin_lock(session, &chunkcache->chunkcache_lru_lock);                 \
+        TAILQ_INSERT_HEAD(&chunkcache->chunkcache_lru_list, chunk, next_lru_item); \
+        __wt_spin_unlock(session, &chunkcache->chunkcache_lru_lock);               \
+    }
+#endif
+
+#define WT_CHUNK_OFFSET(chunkcache, offset) \
+    (wt_off_t)(((size_t)offset / chunkcache->chunk_size) * chunkcache->chunk_size)
+
+#define WT_BUCKET_CHUNKS(chunkcache, bucket_id) &(chunkcache->hashtable[bucket_id].colliding_chunks)
+
+/*
+ * __chunkcache_alloc_space --
+ *     Allocate memory for the chunk in the cache.
+ */
+static int
+__chunkcache_alloc_space(WT_SESSION_IMPL *session, WT_CHUNKCACHE_CHUNK *chunk)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_DECL_RET;
+
+    chunkcache = &S2C(session)->chunkcache;
+
+    if (chunkcache->type == WT_CHUNKCACHE_DRAM)
+        ret = __wt_malloc(session, chunk->chunk_size, &chunk->chunk_location);
+    else {
+#ifdef ENABLE_MEMKIND
+        chunk->chunk_location = memkind_malloc(chunkcache->mem_kind, chunk->chunk_size);
+        if (chunk->chunk_location == NULL)
+            ret = WT_ERROR;
+#else
+        WT_RET_MSG(session, EINVAL,
+          "Chunk cache requires libmemkind, unless it is configured to be in DRAM");
+#endif
+    }
+    if (ret == 0) {
+        __wt_atomic_add64(&chunkcache->bytes_used, chunk->chunk_size);
+        WT_STAT_CONN_INCR(session, chunkcache_chunks);
+        WT_STAT_CONN_INCRV(session, chunkcache_bytes, chunk->chunk_size);
+    }
+    return (ret);
+}
+
+/*
+ * __chunkcache_admit_size --
+ *     Decide if we can admit the chunk given the limit on cache capacity and return the size of the
+ *     chunk to be admitted.
+ */
+static size_t
+__chunkcache_admit_size(WT_SESSION_IMPL *session)
+{
+    WT_CHUNKCACHE *chunkcache;
+
+    chunkcache = &S2C(session)->chunkcache;
+
+    if ((chunkcache->bytes_used + chunkcache->chunk_size) < chunkcache->capacity)
+        return (chunkcache->chunk_size);
+
+    WT_STAT_CONN_INCR(session, chunkcache_exceeded_capacity);
+    __wt_verbose(session, WT_VERB_CHUNKCACHE, "exceeded chunkcache capacity of %" PRIu64 " bytes",
+      chunkcache->capacity);
+    return (0);
+}
+
+/*
+ * __chunkcache_alloc_chunk --
+ *     Allocate the chunk and its metadata for a block at a given offset.
+ */
+static int
+__chunkcache_alloc_chunk(WT_SESSION_IMPL *session, WT_CHUNKCACHE_CHUNK **newchunk, wt_off_t offset,
+  WT_BLOCK *block, WT_CHUNKCACHE_HASHID *hash_id)
```

Comment:
Out parameters should be at the end of the arg list.

Reply:
Done.